torchdynamo/guards.py

```diff
@@ -371,7 +373,9 @@ def compile_check_fn(self, local_builder, global_builder):
         )
         # TODO(whc) maybe only the 'check_tensors' one is ambiguous? if so we can be less general..
         verbose_code_parts = (
-            ["___guarded_code.valid"] + local_builder.code + global_builder.code
+            ["__guarded_code_invalid_reason()"]
```

Comment:
I don't think you need to actually compile anything to do with `reason` into the `check_fn`. I would keep this as `.valid` but then store a .reason alongside where you invalidate.

Reply:
well, this is something we could overhaul in general, but right now the way I did the verbose printing was to construct a duplicate 'check function' which only runs in the verbose log printer and has 1:1 lines with the original check_function.

The original check_function stays 'fast', but if it fails and enters the callback that prints logs, we run the slow/verbose variant which can produce explanation strings.

Reply:
this is pretty hacky.  but how it works is,
1) the fast 'check_fn' fails
2) the callback gets triggered
https://github.com/pytorch/torchdynamo/blob/main/torchdynamo/guards.py#L454
3) the slow 'verbose_check_fn' code parts get iterated, and run one by one
 - True bools or None strings cause iteration to continue, 
 - false bools cause the code part to be printed directly (the legacy fallback)
 - strings returned by verbose codeparts are directly used

Reply:
putting perf considerations aside, it seems like a clean api would be to define a 'guard' obj which includes 2 functions (check() and reason())

the fast path check_fn would consist of something like 
  for g in guards()
       if not g.check():
           FAIL
           
 and then the verbose function could simply run the same loop but instead of doing whatever 'FAIL' means it could call g.reason() and log it.
 
 I went with the hack above to avoid messing with the fast path at all, but it should be rethought.

Reply:
Maybe this is too much for this PR, but instead of having to iterate guards over that failed, we should probably produce guards with a side effect of storing their failure, and then instead of re-iterating and re-evaluating guards, you can just print go and see exactly which guards failed. O(1 * failures) instead of O(guards * failures). This would also allow us to not special case "valid" as each guard could have a "reason" field, with fallback to just printing the failed function as executed at time of failure if no "reason" is provided.

But for this PR, I dont think we need to modify what goes into `check_fn`. You can keep it as `invalidated_reason` where you have it, but just make sure to pass it through as a field and add it to `guard_fn` in this same func (`compile_check_fn`). Then in your guard fail hook, you can just read it off the guard_fn.

Reply:
one reason i didn't go this route (for verbose tensor_check, or for other not yet implemented verbose checks) is that i want to be able to format a string including info from both the original (guarded on) value (such as its size, id, etc), and from the new value (such as its size, id, value, etc).  It's a bit restrictive to have to prepare the string at guard creation time, since at that time you only have access to the original value.

By creating a 'verbose_check_fn' and rerunning its parts at fail-hook time, you can format messages with info from both old and new(runtime) values.  I think this is worth the cost, as we shouldn't care about the performance of our guard-fail-hook.